platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java

```diff
@@ -0,0 +1,375 @@
+
+import static com.swirlds.common.io.utility.TemporaryFileBuilder.buildTemporaryFile;
+import static com.swirlds.logging.LogMarker.STARTUP;
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+
+import com.swirlds.common.config.StateConfig;
+import com.swirlds.common.context.PlatformContext;
+import com.swirlds.common.formatting.TextTable;
+import com.swirlds.common.io.SelfSerializable;
+import com.swirlds.common.io.streams.SerializableDataInputStream;
+import com.swirlds.common.io.streams.SerializableDataOutputStream;
+import com.swirlds.common.io.utility.FileUtils;
+import com.swirlds.common.scratchpad.Scratchpad;
+import com.swirlds.common.scratchpad.ScratchpadType;
+import com.swirlds.common.system.NodeId;
+import com.swirlds.common.threading.locks.AutoClosableLock;
+import com.swirlds.common.threading.locks.Locks;
+import com.swirlds.common.threading.locks.locked.Locked;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A utility for "taking notes" that are preserved across restart boundaries.
+ * <p>
+ * A scratchpad instance is thread safe. All read operations and write operations against a scratchpad are atomic. Any
+ * write that has completed is guaranteed to be visible to all subsequent reads, regardless of crashes/restarts.
+ *
+ * @param <K> the enum type that defines the scratchpad fields
+ */
+public class StandardScratchpad<K extends Enum<K> & ScratchpadType> implements Scratchpad<K> {
+
+    private static final Logger logger = LogManager.getLogger(StandardScratchpad.class);
+
+    /**
+     * The directory where scratchpad files are written. Files are written with the format "0.scr", "1.scr", etc., with
+     * the number increasing by one each time a file is written. If multiple files are present, the file with the
+     * highest number is the most recent and is always used. Multiple files will only be present if the platform crashes
+     * after writing a file but before it has the opportunity to delete the previous file.
+     */
+    public static final String SCRATCHPAD_DIRECTORY_NAME = "scratchpad";
+
+    /**
+     * The file extension used for scratchpad files.
+     */
+    public static final String SCRATCHPAD_FILE_EXTENSION = ".scr";
+
+    private final Set<K> fields;
+    private final Map<Integer, K> indexToFieldMap;
```

Comment:
Only used in constructor - could optionally be made into a local variable

Reply:
It's used by a function, but that function is only called by the constructor, so I agree it shouldn't be a class member variable. Refactored.